File: src/melon/cpp/Platform.cpp
#include <stdio.h>
#include <string>

#include "Platform.h"
#include "Util.cpp"

#include "melon-rs/src/melon/sys.rs.h"

using namespace melonDS;

// this file implements the interface defined by melonDS in Platform.h
// wherever possible, it uses the minimum amount of C++ code to invoke
// the Rust solution

namespace melonDS::Platform
{
    // initialize std::string with ::rust::String
    std::string InstanceFileSuffix()
    {
        return std::string(PlatformImpl::InstanceFileSuffix());
    }

    int InstanceID()
    {
        return PlatformImpl::InstanceID();
    }

    // synchronization primitives
    Thread *Thread_Create(std::function<void()> func)
    {
        Util::OpaqueFunction *opaque = new Util::OpaqueFunction;
        opaque->f = func;
        return PlatformImpl::Thread_Create(opaque);
    }
    void Thread_Free(Thread *thread)
    {
        return PlatformImpl::Thread_Free(thread);
    }
    void Thread_Wait(Thread *thread)
    {
        return PlatformImpl::Thread_Wait(thread);
    }

    Semaphore *Semaphore_Create()
    {
        return PlatformImpl::Semaphore_Create();
    }

    void Semaphore_Free(Semaphore *sema)
    {
        return PlatformImpl::Semaphore_Free(sema);
    }

    void Semaphore_Reset(Semaphore *sema)
    {
        return PlatformImpl::Semaphore_Reset(sema);
    }

    void Semaphore_Wait(Semaphore *sema)
    {
        return PlatformImpl::Semaphore_Wait(sema);
    }

    void Semaphore_Post(Semaphore *sema, int count)
    {
        return PlatformImpl::Semaphore_Post(sema, count);
    }

    Mutex *Mutex_Create()
    {
        return PlatformImpl::Mutex_Create();
    }
    void Mutex_Free(Mutex *mutex)
    {
        return PlatformImpl::Mutex_Free(mutex);
    }
    void Mutex_Lock(Mutex *mutex)
    {
        return PlatformImpl::Mutex_Lock(mutex);
    }
    void Mutex_Unlock(Mutex *mutex)
    {
        return PlatformImpl::Mutex_Unlock(mutex);
    }
    bool Mutex_TryLock(Mutex *mutex)
    {
        return PlatformImpl::Mutex_TryLock(mutex);
    }

    FileHandle *OpenFile(const std::string &path, FileMode mode)
    {
        return PlatformImpl::OpenFile(path, mode);
    }
    FileHandle *OpenLocalFile(const std::string &path, FileMode mode)
    {
        return PlatformImpl::OpenLocalFile(path, mode);
    }

    bool FileExists(const std::string &name)
    {
        return PlatformImpl::FileExists(name);
    }
    bool LocalFileExists(const std::string &name)
    {
        return PlatformImpl::LocalFileExists(name);
    }

    bool CloseFile(FileHandle *file)
    {
        return PlatformImpl::CloseFile(file);
    }

    bool IsEndOfFile(FileHandle *file)
    {
        return PlatformImpl::IsEndOfFile(file);
    }

    bool FileReadLine(char *str, int count, FileHandle *file)
    {
        return PlatformImpl::FileReadLine((u8 *)str, count, file);
    }

    bool FileSeek(FileHandle *file, s64 offset, FileSeekOrigin origin)
    {
        return PlatformImpl::FileSeek(file, offset, origin);
    }

    void FileRewind(FileHandle *file)
    {
        return PlatformImpl::FileRewind(file);
    }

    u64 FileRead(void *data, u64 size, u64 count, FileHandle *file)
    {
        return PlatformImpl::FileRead((u8 *)data, size, count, file);
    }

    bool FileFlush(FileHandle *file)
    {
        return PlatformImpl::FileFlush(file);
    }

    u64 FileWrite(const void *data, u64 size, u64 count, FileHandle *file)
    {
        return PlatformImpl::FileWrite((u8 *)data, size, count, file);
    }

    u64 FileWriteFormatted(FileHandle *file, const char *fmt, ...)
    {
        if (fmt == nullptr)
            return 0;

        va_list args;
        va_start(args, fmt);
        // get required length
        int buf_len = std::snprintf(nullptr, 0, fmt, args);
        if (buf_len <= 0)
        {
            va_end(args);
            return 0;
        }
        auto buf = std::make_unique<char[]>(buf_len);
        // actually write into buffer
        snprintf(buf.get(), buf_len, fmt, args);
        va_end(args);
        return FileWrite((void *)buf.get(), 1, buf_len, file);
    }

    u64 FileLength(FileHandle *file)
    {
        return PlatformImpl::FileLength(file);
    }

    void SignalStop(StopReason reason) {
        // no op
    }
    void Log(LogLevel level, const char* fmt, ...) {
        if (fmt == nullptr)
            return;

        va_list args;
        va_start(args, fmt);
        vprintf(fmt, args);
        va_end(args);
    }
    void WriteFirmware(const Firmware& firmware, u32 writeoffset, u32 writelen) {
        // no op
    }
    void WriteDateTime(int year, int month, int day, int hour, int minute, int second) {
        // no op
    }

    void WriteNDSSave(const u8 *savedata, u32 savelen, u32 writeoffset, u32 writelen)
    {
        return PlatformImpl::WriteNDSSave(savedata, savelen, writeoffset, writelen);
    }
    bool MP_Init()
    {
        return PlatformImpl::MP_Init();
    }
    void MP_DeInit()
    {
        return PlatformImpl::MP_DeInit();
    }
    void MP_Begin()
    {
        return PlatformImpl::MP_Begin();
    }
    void MP_End()
    {
        return PlatformImpl::MP_End();
    }
    int MP_SendPacket(u8 *data, int len, u64 timestamp)
    {
        return PlatformImpl::MP_SendPacket(data, len, timestamp);
    }
    int MP_RecvPacket(u8 *data, u64 *timestamp)
    {
        return PlatformImpl::MP_RecvPacket(data, timestamp);
    }
    int MP_SendCmd(u8 *data, int len, u64 timestamp)
    {
        return PlatformImpl::MP_SendCmd(data, len, timestamp);
    }
    int MP_SendReply(u8 *data, int len, u64 timestamp, u16 aid)
    {
        return PlatformImpl::MP_SendReply(data, len, timestamp, aid);
    }
    int MP_SendAck(u8 *data, int len, u64 timestamp)
    {
        return PlatformImpl::MP_SendAck(data, len, timestamp);
    }
    int MP_RecvHostPacket(u8 *data, u64 *timestamp)
    {
        return PlatformImpl::MP_RecvHostPacket(data, timestamp);
    }
    u16 MP_RecvReplies(u8 *data, u64 timestamp, u16 aidmask)
    {
        return PlatformImpl::MP_RecvReplies(data, timestamp, aidmask);
    }
    bool LAN_Init()
    {
        return PlatformImpl::LAN_Init();
    }
    void LAN_DeInit()
    {
        return PlatformImpl::LAN_DeInit();
    }
    int LAN_SendPacket(u8 *data, int len)
    {
        return PlatformImpl::LAN_SendPacket(data, len);
    }
    int LAN_RecvPacket(u8 *data)
    {
        return PlatformImpl::LAN_RecvPacket(data);
    }
    void Camera_Start(int num)
    {
        return PlatformImpl::Camera_Start(num);
    }
    void Camera_Stop(int num)
    {
        return PlatformImpl::Camera_Stop(num);
    }
    void Camera_CaptureFrame(int num, u32 *frame, int width, int height, bool yuv)
    {
        return PlatformImpl::Camera_CaptureFrame(num, frame, width, height, yuv);
    }
}

File: src/melon/sys/platform.rs
use std::{
    ptr::drop_in_place,
    slice,
    sync::{Mutex, MutexGuard, TryLockError},
    thread::{spawn, JoinHandle},
};

use crate::melon::{save::write_save, subscriptions};

#[cxx::bridge]
pub mod sys {
    #[namespace = "Glue"]
    extern "Rust" {
        #[cxx_name = "Thread"]
        #[namespace = "Platform"]
        type NdsThread;

        #[cxx_name = "Semaphore"]
        #[namespace = "Platform"]
        type NdsSemaphore;

        #[cxx_name = "Mutex"]
        #[namespace = "Platform"]
        type NdsMutex;

        #[cxx_name = "InstanceID"]
        fn instance_id() -> i32;
        #[cxx_name = "InstanceFileSuffix"]
        fn instance_file_suffix() -> String;

        #[cxx_name = "StopEmu"]
        fn stop_emu();

        #[cxx_name = "Camera_Start"]
        fn camera_start(num: i32);
        #[cxx_name = "Camera_Stop"]
        fn camera_stop(num: i32);
        #[cxx_name = "Camera_CaptureFrame"]
        unsafe fn camera_capture_frame(
            num: i32,
            frame: *const u32,
            width: i32,
            height: i32,
            yuv: bool,
        );

        #[cxx_name = "Thread_Create"]
        unsafe fn thread_create(func: *mut OpaqueFunction) -> *mut NdsThread;
        #[cxx_name = "Thread_Wait"]
        unsafe fn thread_wait(thread: *mut NdsThread);
        #[cxx_name = "Thread_Free"]
        unsafe fn thread_free(thread: *mut NdsThread);

        #[cxx_name = "Semaphore_Create"]
        fn semaphore_create() -> *mut NdsSemaphore;
        #[cxx_name = "Semaphore_Free"]
        unsafe fn semaphore_free(sema: *mut NdsSemaphore);
        #[cxx_name = "Semaphore_Reset"]
        unsafe fn semaphore_reset(sema: *mut NdsSemaphore);
        #[cxx_name = "Semaphore_Wait"]
        unsafe fn semaphore_wait(sema: *mut NdsSemaphore);
        #[cxx_name = "Semaphore_Post"]
        unsafe fn semaphore_post(sema: *mut NdsSemaphore, count: i32);

        #[cxx_name = "Mutex_Create"]
        fn mutex_create() -> *mut NdsMutex;
        #[cxx_name = "Mutex_Free"]
        unsafe fn mutex_free(mutex: *mut NdsMutex);
        #[cxx_name = "Mutex_Lock"]
        unsafe fn mutex_lock(mutex: *mut NdsMutex);
        #[cxx_name = "Mutex_TryLock"]
        unsafe fn mutex_try_lock(mutex: *mut NdsMutex) -> bool;
        #[cxx_name = "Mutex_Unlock"]
        unsafe fn mutex_unlock(mutex: *mut NdsMutex);

        #[cxx_name = "LAN_Init"]
        fn lan_init() -> bool;
        #[cxx_name = "LAN_DeInit"]
        fn lan_deinit();
        #[cxx_name = "LAN_SendPacket"]
        unsafe fn lan_send_packet(data: *mut u8, len: i32) -> i32;
        #[cxx_name = "LAN_RecvPacket"]
        unsafe fn lan_recv_packet(data: *mut u8) -> i32;

        // multiplayer
        #[cxx_name = "MP_SendAck"]
        unsafe fn mp_send_ack(data: *mut u8, len: i32, timestamp: u64) -> i32;
        #[cxx_name = "MP_SendCmd"]
        unsafe fn mp_send_cmd(data: *mut u8, len: i32, timestamp: u64) -> i32;
        #[cxx_name = "MP_SendReply"]
        unsafe fn mp_send_reply(data: *mut u8, len: i32, timestamp: u64, aid: u16) -> i32;
        #[cxx_name = "MP_SendPacket"]
        unsafe fn mp_send_packet(data: *mut u8, len: i32, timestamp: u64) -> i32;
        #[cxx_name = "MP_RecvPacket"]
        unsafe fn mp_recv_packet(data: *mut u8, timestamp: *mut u64) -> i32;
        #[cxx_name = "MP_RecvHostPacket"]
        unsafe fn mp_recv_host_packet(data: *mut u8, timestamp: *mut u64) -> i32;
        #[cxx_name = "MP_RecvReplies"]
        unsafe fn mp_recv_replies(data: *mut u8, timestamp: u64, aidmask: u16) -> u16;
        #[cxx_name = "MP_Init"]
        fn mp_init() -> bool;
        #[cxx_name = "MP_DeInit"]
        fn mp_deinit();
        #[cxx_name = "MP_Begin"]
        fn mp_begin();
        #[cxx_name = "MP_End"]
        fn mp_end();

        #[cxx_name = "WriteNDSSave"]
        unsafe fn write_nds_save(
            savedata: *const u8,
            savelen: u32,
            writeoffset: u32,
            writelen: u32,
        );
    }

    #[namespace = "Util"]
    unsafe extern "C++" {
        include!("Util.h");

        type OpaqueFunction;

        unsafe fn OpaqueFunction_Call(func: *mut OpaqueFunction);
        unsafe fn OpaqueFunction_Free(func: *mut OpaqueFunction);
    }
}

fn instance_id() -> i32 {
    // std::process::id() as i32
    0
}

fn instance_file_suffix() -> String {
    format!("{}", instance_id())
}

fn stop_emu() {
    subscriptions::STOP_EMU.lock().unwrap().call(())
}

fn camera_start(num: i32) {}

fn camera_stop(num: i32) {}

unsafe fn camera_capture_frame(
    num: i32,
    frame: *const u32,
    width: i32,
    height: i32,
    yuv: bool,
) {
}

struct NdsThread {
    inner: Option<std::thread::JoinHandle<()>>,
}

impl NdsThread {
    pub fn new(func: JoinHandle<()>) -> Self {
        Self { inner: Some(func) }
    }

    unsafe fn wait(this: *mut Self) {
        (*this).inner.take().unwrap().join().unwrap();
    }
}

use crate::melon::sys::platform::{
    OpaqueFunction, OpaqueFunction_Call, OpaqueFunction_Free,
};

struct OpaqueWrapper(*mut OpaqueFunction);

impl OpaqueWrapper {
    pub unsafe fn run(&mut self) {
        OpaqueFunction_Call(self.0);
    }
}

impl Drop for OpaqueWrapper {
    fn drop(&mut self) {
        unsafe {
            OpaqueFunction_Free(self.0);
        }
    }
}

unsafe impl Send for OpaqueWrapper {}
unsafe impl Sync for OpaqueWrapper {}

unsafe fn thread_create(func: *mut OpaqueFunction) -> *mut NdsThread {
    let mut wrapper = OpaqueWrapper(func);
    let nds_thread = Box::new(NdsThread::new(spawn(move || {
        wrapper.run();
    })));
    Box::leak(nds_thread)
}
unsafe fn thread_wait(thread: *mut NdsThread) {
    NdsThread::wait(thread);
}
unsafe fn thread_free(thread: *mut NdsThread) {
    drop_in_place(thread);
}

struct NdsSemaphore {
    capacity: Mutex<usize>,
}

impl NdsSemaphore {
    pub fn new() -> Self {
        Self {
            capacity: Mutex::new(0),
        }
    }
}

fn semaphore_create() -> *mut NdsSemaphore {
    let sema = Box::new(NdsSemaphore::new());
    Box::leak(sema)
}
// drops the semaphore
unsafe fn semaphore_free(sema: *mut NdsSemaphore) {
    drop_in_place(sema);
}
// acquire all available resources
unsafe fn semaphore_reset(sema: *mut NdsSemaphore) {
    *(*sema).capacity.get_mut().unwrap() = 0;
}
// get one resource
unsafe fn semaphore_wait(sema: *mut NdsSemaphore) {
    loop {
        let available = (*sema).capacity.get_mut().unwrap();
        if *available > 0 {
            *available -= 1;
            break;
        }
    }
}
// release a certain number of resources
// memory stuff is managed in cpp. Just add them I guess
unsafe fn semaphore_post(sema: *mut NdsSemaphore, count: i32) {
    let cap = (*sema).capacity.get_mut().unwrap();
    *cap += count as usize;
}

// probably invoking some 8th cardinal sin
struct NdsMutex {
    mutex: Mutex<()>,
    guard: Option<MutexGuard<'static, ()>>,
}

impl NdsMutex {
    pub fn new() -> NdsMutex {
        Self {
            mutex: Mutex::new(()),
            guard: None,
        }
    }

    pub unsafe fn lock(this: *mut Self) {
        let guard = (*this).mutex.lock().unwrap();
        (*this).guard = Some(guard);
    }

    pub unsafe fn try_lock(
        this: *mut Self,
    ) -> Result<(), TryLockError<MutexGuard<'static, ()>>> {
        let guard = (*this).mutex.try_lock()?;
        (*this).guard = Some(guard);
        Ok(())
    }

    pub unsafe fn unlock(this: *mut Self) {
        (*this).guard.take();
    }
}

fn mutex_create() -> *mut NdsMutex {
    let mutex = Box::new(NdsMutex::new());
    Box::leak(mutex)
}
unsafe fn mutex_lock(mutex: *mut NdsMutex) {
    NdsMutex::lock(mutex);
}
unsafe fn mutex_try_lock(mutex: *mut NdsMutex) -> bool {
    NdsMutex::try_lock(mutex).is_ok()
}
unsafe fn mutex_unlock(mutex: *mut NdsMutex) {
    NdsMutex::unlock(mutex);
}
unsafe fn mutex_free(mutex: *mut NdsMutex) {
    drop_in_place(mutex);
}

fn lan_init() -> bool {
    // TODO: provide an event subscription
    false
}

unsafe fn lan_send_packet(data: *mut u8, len: i32) -> i32 {
    0
}
unsafe fn lan_recv_packet(data: *mut u8) -> i32 {
    0
}

fn lan_deinit() {
    subscriptions::STOP_EMU.lock().unwrap().call(())
}

unsafe fn mp_send_ack(data: *mut u8, len: i32, timestamp: u64) -> i32 {
    0
}
unsafe fn mp_send_cmd(data: *mut u8, len: i32, timestamp: u64) -> i32 {
    0
}
unsafe fn mp_send_reply(data: *mut u8, len: i32, timestamp: u64, aid: u16) -> i32 {
    0
}
unsafe fn mp_send_packet(data: *mut u8, len: i32, timestamp: u64) -> i32 {
    0
}
unsafe fn mp_recv_packet(data: *mut u8, timestamp: *mut u64) -> i32 {
    0
}
unsafe fn mp_recv_host_packet(data: *mut u8, timestamp: *mut u64) -> i32 {
    0
}
unsafe fn mp_recv_replies(data: *mut u8, timestamp: u64, aidmask: u16) -> u16 {
    0
}
fn mp_init() -> bool {
    true
}
fn mp_deinit() {}
fn mp_begin() {}
fn mp_end() {}

unsafe fn write_nds_save(savedata: *const u8, savelen: u32, writeoffset: u32, writelen: u32) {
    write_save(
        slice::from_raw_parts(savedata, savelen as usize),
        writeoffset as usize,
        writelen as usize,
    );
}

pub use sys::*;
